File: android/core/realtime/src/main/kotlin/com/dualmusic/core/realtime/RealtimeClient.kt
package com.dualmusic.core.realtime

import com.dualmusic.domain.realtime.Realtime
import io.socket.client.IO
import io.socket.client.Socket
import kotlinx.serialization.DeserializationStrategy
import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer
import org.json.JSONObject
import java.net.URI

/**
 * Client temps réel Socket.IO de Dual Music (Android).
 *
 * Un [Socket] par namespace (`/chat`, `/live`, `/notifications`), partageant le handshake
 * JWT. Reconnexion exponentielle (1→30 s) déléguée au SDK. Les événements serveur sont
 * exposés via [NamespaceSession.on] avec décodage typé kotlinx.serialization.
 *
 * Exemple :
 * ```
 * val rt = RealtimeClient(baseUrl) { tokenStore.accessToken() }
 * val live = rt.session(Realtime.Namespace.LIVE)
 * live.onConnect { live.join(Realtime.RoomType.DUEL, duelId) }
 * val sub = live.on(Realtime.RealtimeEvent.VOTE, VotePayload.serializer()) { vote -> ... }
 * live.connect()
 * ```
 *
 * @param baseUrl URL publique du backend (ex. `https://api.dualmusic.app`).
 * @param jwtProvider fournit le JWT courant pour le handshake.
 */
class RealtimeClient(
    private val baseUrl: String,
    private val jwtProvider: suspend () -> String?,
) {
    private val sessions = mutableMapOf<String, NamespaceSession>()
    private val json = Json { ignoreUnknownKeys = true; explicitNulls = false }

    /** Récupère (ou crée) la session d'un namespace. */
    @Synchronized
    fun session(namespace: String): NamespaceSession =
        sessions.getOrPut(namespace) { NamespaceSession(baseUrl, namespace, jwtProvider, json) }

    /** Déconnecte tous les namespaces (arrière-plan / logout). */
    fun disconnectAll() = sessions.values.forEach { it.disconnect() }
}

/** Session d'un namespace : connexion, join/leave, écoute typée. */
class NamespaceSession internal constructor(
    private val baseUrl: String,
    private val namespace: String,
    private val jwtProvider: suspend () -> String?,
    private val json: Json,
) {
    private var socket: Socket? = null

    // Handlers mémorisés AVANT que le socket existe. Indispensable : le VM appelle
    // onConnect{}/on{} AVANT connect(), donc `socket` est null à ce moment-là. On les
    // (ré)enregistre sur le socket dans connect() (et à chaque reconnexion/re-création),
    // sinon les callbacks sont perdus (bug: le /chat ne rejoignait jamais sa room).
    private val connectHandlers = mutableListOf<() -> Unit>()
    private val disconnectHandlers = mutableListOf<() -> Unit>()
    private val eventHandlers = mutableListOf<Pair<String, (Array<out Any?>) -> Unit>>()

    /** Construit les options avec le JWT en handshake (`auth.token`) + reconnexion. */
    private fun options(token: String?): IO.Options = IO.Options.builder()
        .setTransports(arrayOf("websocket"))
        .setReconnection(true)
        .setReconnectionDelay(1_000)
        .setReconnectionDelayMax(30_000)
        .setRandomizationFactor(0.5) // jitter
        .setAuth(token?.let { mapOf("token" to it) } ?: emptyMap())
        .build()

    /** Connecte le socket du namespace après injection du JWT + (ré)applique tous les handlers. */
    suspend fun connect() {
        val token = jwtProvider()
        val uri = URI.create(baseUrl.trimEnd('/') + namespace)
        socket?.disconnect() // ferme l'ancien socket avant d'en créer un nouveau (pas de doublon)
        val s = IO.socket(uri, options(token))
        // (Ré)enregistre TOUS les handlers mémorisés sur ce nouveau socket, peu importe l'ordre
        // des appels onConnect/on par rapport à connect().
        connectHandlers.forEach { h -> s.on(Socket.EVENT_CONNECT) { h() } }
        disconnectHandlers.forEach { h -> s.on(Socket.EVENT_DISCONNECT) { h() } }
        eventHandlers.forEach { (event, raw) -> s.on(event) { args -> raw(args) } }
        socket = s
        s.connect()
    }

    /** Déconnecte + oublie les handlers (session partagée : évite l'accumulation entre écrans). */
    fun disconnect() {
        socket?.disconnect()
        socket = null
        connectHandlers.clear()
        disconnectHandlers.clear()
        eventHandlers.clear()
    }

    /** Rejoint une room `type:id`. */
    fun join(type: Realtime.RoomType, id: String) {
        socket?.emit(Realtime.EVENT_JOIN, JSONObject(mapOf("type" to type.wire, "id" to id)))
    }

    /** Quitte une room. */
    fun leave(type: Realtime.RoomType, id: String) {
        socket?.emit(Realtime.EVENT_LEAVE, JSONObject(mapOf("type" to type.wire, "id" to id)))
    }

    /**
     * Émet un événement arbitraire (client → serveur). Utilisé pour le relais de broadcast
     * éphémère (`broadcast:join`, `broadcast`) : réactions emojis, animations… Le serveur
     * relaie aux autres membres du canal (l'émetteur applique son effet localement).
     * @param payload argument JSON (String de canal, ou [JSONObject]). `null` = sans argument.
     */
    fun emit(event: String, payload: Any? = null) {
        if (payload == null) socket?.emit(event) else socket?.emit(event, payload)
    }

    /** Callback de connexion établie (pour (re)join les rooms). Mémorisé + posé si socket prêt. */
    fun onConnect(handler: () -> Unit) {
        connectHandlers += handler
        socket?.on(Socket.EVENT_CONNECT) { handler() }
    }

    /** Callback de déconnexion. Mémorisé + posé si socket prêt. */
    fun onDisconnect(handler: () -> Unit) {
        disconnectHandlers += handler
        socket?.on(Socket.EVENT_DISCONNECT) { handler() }
    }

    /**
     * Écoute un événement serveur et décode son premier argument JSON en [T].
     * Le handler est mémorisé (et réappliqué à chaque connect) pour survivre à l'ordre des appels.
     * @param event nom d'événement (voir [Realtime.RealtimeEvent]).
     * @param deserializer serializer kotlinx du payload.
     */
    fun <T> on(event: String, deserializer: DeserializationStrategy<T>, handler: (T) -> Unit) {
        val raw: (Array<out Any?>) -> Unit = { args ->
            val first = args.firstOrNull() as? JSONObject
            if (first != null) {
                runCatching { json.decodeFromString(deserializer, first.toString()) }
                    .getOrNull()
                    ?.let(handler)
            }
        }
        eventHandlers += event to raw
        socket?.on(event) { args -> raw(args) }
    }

    /** Variante reified : `session.on<VotePayload>(VOTE) { ... }`. */
    inline fun <reified T> on(event: String, noinline handler: (T) -> Unit) =
        on(event, serializer(), handler)
}

File: ios/Packages/CoreRealtime/Sources/CoreRealtime/RealtimeClient.swift
import Foundation
import SocketIO

/// Client temps réel Socket.IO de Dual Music (iOS).
///
/// Un `SocketManager` par URL de base ; chaque namespace (`/chat`, `/live`,
/// `/notifications`) est un `SocketIOClient` distinct partageant le handshake JWT.
/// Reconnexion exponentielle gérée par le SDK (1→30 s). Les événements serveur sont
/// exposés en handlers typés décodés depuis JSON.
///
/// Cycle de vie type :
/// ```
/// let rt = RealtimeClient(baseURL: url) { await tokenStore.accessToken() }
/// let live = rt.session(.live)
/// live.onConnect { live.join(.duel, id: duelId) }
/// let sub = live.onEvent(Realtime.Event.vote, as: VotePayload.self) { vote in ... }
/// live.connect()
/// ```
public final class RealtimeClient {

    private let manager: SocketManager
    private var sessions: [Realtime.Namespace: NamespaceSession] = [:]
    private let lock = NSLock()

    /// - Parameters:
    ///   - baseURL: URL publique du backend (ex. `https://api.dualmusic.app`). Le SDK gère
    ///     l'upgrade en WebSocket sur `/socket.io/`.
    ///   - jwtProvider: fournit le JWT courant pour le handshake (connectParams `token`).
    public init(baseURL: URL, jwtProvider: @escaping @Sendable () async -> String?) {
        // Note : le token est résolu de façon synchrone au moment de la config. Pour un
        // refresh, appeler `reconnect(with:)` avec le nouveau jeton.
        self.manager = SocketManager(
            socketURL: baseURL,
            config: [
                .log(false),
                .compress,
                .reconnects(true),
                .reconnectWait(1),
                .reconnectWaitMax(30),
                .randomizationFactor(0.5), // jitter
                .forceWebsockets(true),
            ]
        )
        self.jwtProvider = jwtProvider
    }

    private let jwtProvider: @Sendable () async -> String?

    /// Récupère (ou crée) la session d'un namespace.
    public func session(_ namespace: Realtime.Namespace) -> NamespaceSession {
        lock.lock(); defer { lock.unlock() }
        if let existing = sessions[namespace] { return existing }
        let socket = manager.socket(forNamespace: namespace.rawValue)
        let session = NamespaceSession(namespace: namespace, socket: socket, jwtProvider: jwtProvider)
        sessions[namespace] = session
        return session
    }

    /// Reconnecte tous les namespaces avec un nouveau jeton (après refresh JWT).
    public func reconnect(with token: String) {
        manager.setConfigs([.connectParams(["token": token])])
        manager.reconnect()
    }

    /// Déconnecte tous les namespaces (mise en arrière-plan, logout).
    public func disconnectAll() {
        lock.lock(); let all = sessions.values; lock.unlock()
        all.forEach { $0.disconnect() }
    }
}

/// Session d'un namespace : connexion, join/leave de rooms, écoute typée d'événements.
public final class NamespaceSession {

    private let namespace: Realtime.Namespace
    private let socket: SocketIOClient
    private let jwtProvider: @Sendable () async -> String?
    private let decoder = JSONDecoder()

    init(namespace: Realtime.Namespace, socket: SocketIOClient, jwtProvider: @escaping @Sendable () async -> String?) {
        self.namespace = namespace
        self.socket = socket
        self.jwtProvider = jwtProvider
    }

    /// Connecte le socket après avoir injecté le JWT dans le handshake.
    public func connect() {
        Task {
            let token = await jwtProvider()
            // Le token est passé en connectParams du manager (partagé) ; ici on garantit
            // qu'il est à jour avant le connect du namespace.
            if let token { socket.manager?.setConfigs([.connectParams(["token": token])]) }
            socket.connect()
        }
    }

    public func disconnect() { socket.disconnect() }

    /// Rejoint une room `type:id` (émet `join { type, id }`).
    public func join(_ type: Realtime.RoomType, id: String) {
        socket.emit(Realtime.eventJoin, ["type": type.rawValue, "id": id])
    }

    /// Quitte une room.
    public func leave(_ type: Realtime.RoomType, id: String) {
        socket.emit(Realtime.eventLeave, ["type": type.rawValue, "id": id])
    }

    /// Callback de connexion établie (utile pour (re)join les rooms après reconnexion).
    @discardableResult
    public func onConnect(_ handler: @escaping @Sendable () -> Void) -> Subscription {
        let uuid = socket.on(clientEvent: .connect) { _, _ in handler() }
        return Subscription { [weak socket] in socket?.off(id: uuid) }
    }

    /// Callback de déconnexion.
    @discardableResult
    public func onDisconnect(_ handler: @escaping @Sendable () -> Void) -> Subscription {
        let uuid = socket.on(clientEvent: .disconnect) { _, _ in handler() }
        return Subscription { [weak socket] in socket?.off(id: uuid) }
    }

    /// Écoute un événement serveur et décode son premier argument JSON en `T`.
    /// - Returns: une `Subscription` à conserver ; `cancel()` retire le listener.
    @discardableResult
    public func onEvent<T: Decodable>(_ event: String, as type: T.Type, _ handler: @escaping @Sendable (T) -> Void) -> Subscription {
        let uuid = socket.on(event) { [decoder] data, _ in
            guard let first = data.first else { return }
            guard let payload = try? JSONSerialization.data(withJSONObject: first),
                  let decoded = try? decoder.decode(T.self, from: payload) else { return }
            handler(decoded)
        }
        return Subscription { [weak socket] in socket?.off(id: uuid) }
    }
}

/// Handle d'annulation d'un listener. À conserver tant que l'écoute doit vivre.
public final class Subscription {
    private let onCancel: () -> Void
    private var cancelled = false
    init(_ onCancel: @escaping () -> Void) { self.onCancel = onCancel }
    public func cancel() { guard !cancelled else { return }; cancelled = true; onCancel() }
    deinit { cancel() }
}
